average_neon.c

```diff
@@ -0,0 +1,93 @@
+#include <stdio.h> 
+#include <stdlib.h>
+#include <stdint.h>
+#include <string.h>
+#include <sys/time.h>
+#include <arm_neon.h>
+
+
+//Take the number 1, treat it as long integer and shift its binary representation 
+//30 positions to the left => N = 2^32
+#define N (1L << 30)
+
+
+int main() {
+
+    //allocate memory for array
+    //A array of size N
+	float *A = (float *)malloc(N * sizeof(float));
+	if (!A) {
+		fprintf(stderr, "error allocating %ld bytes\n", N * sizeof(float));
+		exit(1);
+	}
+
+	//fill array with numbers
+	for (int i=0; i < N; i++) {
+		A[i] = (float) i;
+	}
+
+	float avg1 = 0.0, avg2=0.0;
+    struct timeval tv1, tv2, tv3, diff1, diff2;
+	gettimeofday(&tv1, NULL);
+
+
+	//SCALAR VERSION
+	//calculates average in avg1 variable
+	for (int i=0; i < N; i++) {
+		avg1 += A[i];
+	}
+	avg1 /= (float)N;
+	gettimeofday(&tv2, NULL);
+
+
+
+    //NEON VERSION (george mermigkis)
+    //declare vector of 4 floats values and initialize with 0
+    //basically it duplicates float val 0 and assigns it to all 4 parts of vector
+    float32x4_t vec = vdupq_n_f32(0.0f);
+
+    //do N/4 iterations because every NEON vector will take 4 loops each time
+    for(int i=0; i<N; i+=4){
+
+        //load the corresponding 4 values from the loop, perform addition & store result in vec
+        float32x4_t input = vld1q_f32(&A[i]);
+        vec=vaddq_f32(vec,input);
+    }
+    
+    
+    //declare at as 32x2 because it will store only 2 floats after the addition takes place
+    //extract the 2 upper floats and the 2 lower floats and add them
+    float32x2_t upper_half=vadd_f32(vget_high_f32(vec), vget_low_f32(vec));
```

Comment:
very nicely done, but I will suggest you look at the intrinsic vaddvq_f32() and see the difference.

Reply:
Thank you, I will take a look. I'm already making some corrections.